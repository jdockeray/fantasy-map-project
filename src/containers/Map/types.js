// EDIT MODE: Constants
export const ADD = 'ADD'
export const DELETE = 'DELETE'
