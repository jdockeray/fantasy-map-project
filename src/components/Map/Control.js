import React from 'react'
import { Button, ButtonGroup } from 'react-bootstrap'
import { noop } from 'lodash'
import PropTypes from 'prop-types'
import { If, Then, Else } from 'react-if'

export const ControlPropTypes = {
  callback: PropTypes.func.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    text: PropTypes.string,
  })).isRequired,
}

export default function Control({
  callback = noop,
  items = [],
}) {
  return (
    <ButtonGroup className="mapControls">
      {
        items.map(({ label, text, active }) => (


          <If condition={active}>
            <Then>
              <Button
                key={label}
                value={label}
                onClick={callback}
                bsStyle={active && 'primary'}
              >
                { text }
              </Button>
            </Then>
            <Else>{() => (
              <Button
                key={label}
                value={label}
                onClick={callback}
              >
                { text }
              </Button>)
            }
            </Else>
          </If>
        ))
      }
    </ButtonGroup>)
}

Control.propTypes = ControlPropTypes
