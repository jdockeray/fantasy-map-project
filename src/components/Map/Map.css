/* #map {
  border: 1px solid #333;
} */

.mapControls {
  position: fixed;
  z-index: 1000;
}
