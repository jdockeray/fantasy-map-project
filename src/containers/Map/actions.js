import { ADD_MAP_ITEM } from '../../reducers/map'

export const addMapItem = item => ({
  type: ADD_MAP_ITEM,
  payload: {
    item,
  },
})

// Need to flesh this out
export const deleteMapItem = item => ({
  type: ADD_MAP_ITEM,
  payload: {
    item,
  },
})
