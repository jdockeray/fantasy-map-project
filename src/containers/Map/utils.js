
export const isItemActive = (x, y, item = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
}) => {
  const upperLeftCorner = {
    x: item.x,
    y: item.y,
  }

  const lowerRightCorner = {
    x: item.x + (item.width),
    y: item.y + (item.height),
  }
  if (
    (x >= upperLeftCorner.x && x <= lowerRightCorner.x)
    && (y >= upperLeftCorner.y && y <= lowerRightCorner.y)
  ) {
    return true
  }
  return false
}

export const isCursorActive = (x, y, items) =>
  items.reduce((acc, item) => {
    if (acc) return true

    return isItemActive(x, y, item)
  }, false)

export const deselectItems = items => items.map(i => ({ ...i, selected: false }))

export const selectItems = (x, y, items) => {
  if (isCursorActive(x, y, items)) {
    return items.map(item => ({ ...item, selected: item.selected || isItemActive(x, y, item) }))
  }

  return deselectItems(items)
}


export const deleteItems = (x, y, items) => items.filter(item => !isItemActive(x, y, item))

export const addMapItems = (item, items) => [...items, item]
