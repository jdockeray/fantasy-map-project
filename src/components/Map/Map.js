import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { get } from 'lodash'
import './Map.css'
const mapStateToProps = state => ({
  background: get(state, 'form.settings.values.background') || null,
})

const mapDispatchToProps = dispatch => bindActionCreators({
  // myAction
}, dispatch)

class Map extends Component {
  componentDidMount() {
    const {
      canvas,
      image,
    } = this
    const ctx = canvas.getContext('2d')
    // const img = this.refs.image
    // img.onload = () => {
    //   ctx.drawImage(img, 0, 0)
    //   ctx.font = '40px Courier'
    //   ctx.fillText(this.props.text, 210, 75)
    // }
  }
  render() {
    const {
      background
    } = this.props

    return (
      <div>
        <canvas
          ref={(c) => { this.canvas = c }}
          width={640}
          height={425}
          id="map"
          style={{
            backgroundImage: `url(${background})`,
          }}
        />
        {/* <img ref="image" src={cheese} className="hidden" /> */}
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Map)
