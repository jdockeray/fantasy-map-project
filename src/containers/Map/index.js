import React, { Component } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { Grid, Row, Col } from 'react-bootstrap'
import PropTypes from 'prop-types'
import { get } from 'lodash'
import Control from '../../components/Map/Control'
import Map from '../../components/Map/Map'
import { landscapeProps } from '../../helpers/propTypes'
import { addMapItem, deleteMapItems, changeEditMode } from './actions'
import * as editTypes from './types'

const mapStateToProps = state => ({
  background: get(state, 'form.settings.values.background') || null,
  landscape: get(state, 'form.settings.values.landscape') || null,
  map: state.map,
  mapEditMode: state.mapEditMode,
})

const mapDispatchToProps = dispatch => bindActionCreators({
  addMapItem,
  deleteMapItems,
  changeEditMode,
}, dispatch)

class MapWrapper extends Component {
  static propTypes = {
    // actions
    addMapItem: PropTypes.func.isRequired,
    deleteMapItems: PropTypes.func.isRequired,
    changeEditMode: PropTypes.func.isRequired,

    // state
    background: PropTypes.string,
    landscape: landscapeProps,
  }

  defaultProps: {
    background: '',
    landscape: ''
  }

  handleEditMode = (evt) => {
    const editType = get(evt, 'target.value')
    if (Object.keys(editTypes).indexOf(editType) !== -1) {
      this.props.changeEditMode(get(evt, 'target.value'))
    }
  }

  formatTypeForControl = type => ({
    label: type,
    text: type.toLowerCase(),
  })


  render() {
    return (
      <Grid>
        <Row>
          <Control
            items={[
              this.formatTypeForControl(editTypes.ADD),
              this.formatTypeForControl(editTypes.DELETE),
              this.formatTypeForControl(editTypes.SELECT),
            ]}
            callback={this.handleEditMode}
          />
        </Row>
        <Row>
          <Col md={1} />

          <Col md={10} >
            <Map
              {...this.props}
            />
          </Col>
        </Row>
      </Grid>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(MapWrapper)
