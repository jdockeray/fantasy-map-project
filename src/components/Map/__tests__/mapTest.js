import renderer from 'react-test-renderer'

import React from 'react'
import Map from '../Map'

it('renders empty props correctly', () => {
  const mapProps = {
    position: {
      x: 123,
      y: 124,
    },
    landscape: null,
    items: [],
    background: null,
    addMapItem: () => null,
  }

  const map = renderer
    .create(<Map {...mapProps} />)
    .toJSON()

  expect(map).toMatchSnapshot()
})
