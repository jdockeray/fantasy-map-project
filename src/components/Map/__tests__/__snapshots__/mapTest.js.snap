// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty props correctly 1`] = `
<div
  className="mapFrame"
>
  <canvas
    height={600}
    id="map"
    onClick={[Function]}
    style={
      Object {
        "backgroundImage": "url(null)",
        "cursor": "url(null),auto",
      }
    }
    width={900}
  />
</div>
`;
